cmd/soroban-rpc/internal/ledgerentry_storage/ledgerentry_storage.go

```diff
@@ -0,0 +1,238 @@
+package ledgerentry_storage
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"sync"
+	"time"
+
+	"github.com/stellar/go/historyarchive"
+	"github.com/stellar/go/ingest"
+	backends "github.com/stellar/go/ingest/ledgerbackend"
+	"github.com/stellar/go/support/log"
+	"github.com/stellar/go/xdr"
+)
+
+// TODO: Make this configurable?
+const maxBatchSize = 150
+
+type LedgerEntryStorage interface {
+	GetLedgerEntry(key xdr.LedgerKey) (xdr.LedgerEntry, bool, uint32, error)
+	io.Closer
+}
+
+func NewLedgerEntryStorage(logger *log.Entry, db DB, networkPassPhrase string, archive historyarchive.ArchiveInterface, ledgerBackend backends.LedgerBackend) (LedgerEntryStorage, error) {
+	ctx, done := context.WithCancel(context.Background())
+	ls := ledgerEntryStorage{
+		logger:            logger,
+		db:                db,
+		networkPassPhrase: networkPassPhrase,
+		done:              done,
+	}
+	ls.wg.Add(1)
+	go ls.run(ctx, archive, ledgerBackend)
+	return &ls, nil
+}
+
+type ledgerEntryStorage struct {
+	logger            *log.Entry
+	db                DB
+	networkPassPhrase string
+	done              context.CancelFunc
+	wg                sync.WaitGroup
+}
+
+func (ls *ledgerEntryStorage) GetLedgerEntry(key xdr.LedgerKey) (xdr.LedgerEntry, bool, uint32, error) {
+	return ls.db.GetLedgerEntry(key)
+}
+
+func (ls *ledgerEntryStorage) Close() error {
+	ls.done()
+	ls.wg.Wait()
+	ls.db.Close()
+	return nil
+}
+
+func (ls *ledgerEntryStorage) fillEntriesFromLatestCheckpoint(ctx context.Context, archive historyarchive.ArchiveInterface) (uint32, error) {
+	root, err := archive.GetRootHAS()
+	if err != nil {
+		return 0, err
+	}
+	startCheckpointLedger := root.CurrentLedger
+
+	ls.logger.Infof("Starting processing of checkpoint %d", startCheckpointLedger)
+	// TODO: should we make this configurable?
+	checkpointCtx, cancelCheckpointCtx := context.WithTimeout(ctx, 30*time.Minute)
+	defer cancelCheckpointCtx()
+	reader, err := ingest.NewCheckpointChangeReader(checkpointCtx, archive, startCheckpointLedger)
+	if err != nil {
+		return 0, err
+	}
+	tx, err := ls.db.NewLedgerEntryUpdaterTx(startCheckpointLedger, maxBatchSize)
+	if err != nil {
+		return 0, err
+	}
+	// Make sure we finish the updating transaction
+	entryCount := 0
+
+	for {
+		select {
+		case <-ctx.Done():
+			cancelCheckpointCtx()
+			return 0, context.Canceled
+		default:
+		}
+		change, err := reader.Read()
+		if err == io.EOF {
+			break
+		}
+		if err != nil {
+			// TODO: we probably shouldn't panic, at least in case of timeout
+			panic(err)
```

Comment:
no need to panic here. Just return the error and let the caller decide what to do next.

Reply:
👍